Add stm32h7 and stm32mp1 support for spi with dma

shlomow opened the pull request:
This PR adds support for stm32h7 and stm32mp1 socs for spi with dma.
See the commits for the changes that were needed.
In addition, we change spi_loopback test to work with non cacheable memories.
This will allow the tests to be passed for h7/f7 socs until the driver will support cache maintenance operations.

depends on: #33863

shlomow commented:
@FRASTM fyi

arnopo commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -38,12 +38,27 @@ struct spi_cs_control spi_cs = {
+static __aligned(32) uint8_t buffer_tx[BUF_SIZE] __used
+	__attribute__((__section__(".nocache")));
+static const uint8_t BUFFER_TX[] = "0123456789abcdef\0";
+static __aligned(32) uint8_t buffer_rx[BUF_SIZE] __used
```
what about define the const out of the #if so for both configs. and not initialize the buffer2_tx in the #else

arnopo commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -585,6 +600,11 @@ void test_spi_loopback(void)
+#ifdef CONFIG_NOCACHE_MEMORY
+	memcpy(buffer_tx, BUFFER_TX, sizeof(BUFFER_TX));
+	memcpy(buffer2_tx, BUFFER2_TX, sizeof(BUFFER2_TX));
+#endif
```
Then here you can suppress ifdef

FRASTM commented:
Do you have the corresponding test tests/drivers/spi/spi_loopback for one of each board the and stm32mp1 board ?

shlomow commented:
> Do you have the corresponding tests/drivers/spi/spi_loopback for one of each board the and stm32mp1 board ?

I tested that only on `nucleo_h743zi` board:

```
*** Booting Zephyr OS build zephyr-v2.5.0-2280-gd75907b813c8  ***
Running test suite test_spi
===================================================================
START - test_spi_loopback
I: SPI test on buffers TX/RX 0x24000000/0x24000060
I: SPI test slow config
I: Start complete multiple
I: CS control inhibited (no GPIO device)
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Skip half end
I: Skip every 4
I: SPI test fast config
I: Start complete multiple
I: CS control inhibited (no GPIO device)
I: Passed
I: Start complete loop
I: Passed
I: Start null tx
I: Passed
I: Start half start
I: Passed
I: Skip half end
I: Skip every 4
I: Start complete loop
I: CS control inhibited (no GPIO device)
I: Passed
I: Start complete loop
I: CS control inhibited (no GPIO device)
I: Passed
I: All tx/rx passed
 PASS - test_spi_loopback in 0.66 seconds
===================================================================
Test suite test_spi succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

FRASTM reviewed (approved):
stm32h7 and stm32mp1 have SPI with 4-bit to 32-bit data size selection. Could the SPI DMA be  optimized for this large data size ?

arnopo commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -201,4 +201,26 @@ static inline void ll_func_disable_spi(SPI_TypeDef *spi)
+{
+#if defined(CONFIG_SOC_SERIES_STM32MP1X) || \
+	defined(CONFIG_SOC_SERIES_STM32H7X)
+	return (uint32_t) &(spi->TXDR);
```
I wonder if it would be better to define some inline functions in stm32 LL to get Rx and Tx device address instead of accessing directly to the SPI structure.
@erwango , : any preference?

erwango replied:
@arnopo we generally avoid modifying Cube APIs, so I think the proposal is the best that can be done.

arnopo reviewed (commented):
LGTM, with few minor remarks

shlomow committed:
tests: spi: add support for nocache memory

Some devices has cache which may cause problem
when using the test of spi with dma.
This commit allows the spi test buffers to be
places in a non cacheable memory so drivers
that needs to be tested with scenario above will
pass the test. This is not relevant for drivers
that are doing cache maintenance operations.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

shlomow committed:
drivers: spi: stm32: add support for h7 and mp1 with dma

This commit fills the gap for h7 and mp1 socs when
using the spi driver with dma.
These socs have different registers for rx and tx so
setting the dma source and dest is different.
In addition, some ll functions were used only for the
other socs in the dma section. Thus, we replace the ll
functions correctly.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

shlomow committed:
tests: spi: add nucleo_h743zi board

Add nucleo_h743zi to spi_loopback test.
It sets the dma bindings and config the test
to run without cache.

Signed-off-by: Shlomi Vaknin <shlomi.39sd@gmail.com>

shlomow commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -745,7 +760,7 @@ static int transceive_dma(const struct device *dev,
 		spi_context_update_rx(&data->ctx, 1, dma_len);
 	}
 
-	LL_SPI_Disable(spi);
+	/* spi device gets disabled in spi_stm32_complete */
```
@erwango Sorry I am not using zephyr so much at the moment, so I cannot make too much effort in contributing code right now. Sorry again
